pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -243,6 +411,210 @@ func (ctrl *PersistentVolumeController) isFullySynced() bool {
+	if !ok {
+		return nil, fmt.Errorf("Unexpected volume cast error : %v", volumeClone)
+	}
+
+	volumeClone.Status.Phase = phase
+	newVol, err := ctrl.kubeClient.Core().PersistentVolumes().UpdateStatus(volumeClone)
+	if err != nil {
+		glog.V(5).Infof("Updating PersistentVolume[%s]: set phase %s failed: %v", volume.Name, phase, err)
+		return newVol, err
+	}
+	glog.V(5).Infof("Updating PersistentVolume[%s]: phase %s set", volume.Name, phase)
+	return newVol, err
+}
+
+func (ctrl *PersistentVolumeController) bindVolumeToClaim(volume *api.PersistentVolume, claim *api.PersistentVolumeClaim) (*api.PersistentVolume, error) {
+	glog.V(5).Infof("Updating PersistentVolume[%s]: binding to %q", volume.Name, claimToClaimKey(claim))
+
+	// Check if the volume was already bound
+	boundByController := false
+	if !isVolumeBoundToClaim(volume, claim) {
+		boundByController = true
+	}
+
+	// Save the PV only when any modification is neccesary.
+	if volume.Spec.ClaimRef == nil || volume.Spec.ClaimRef.Namespace != claim.Namespace ||
+		volume.Spec.ClaimRef.Name != claim.Name ||
+		volume.Spec.ClaimRef.UID != claim.UID ||
+		(boundByController && !ctrl.hasAnnotation(volume.ObjectMeta, annBoundByController)) {
+
+		clone, err := conversion.NewCloner().DeepCopy(volume)
+		if err != nil {
+			return nil, fmt.Errorf("Error cloning pv: %v", err)
+		}
+		volumeClone, ok := clone.(*api.PersistentVolume)
+		if !ok {
+			return nil, fmt.Errorf("Unexpected volume cast error : %v", volumeClone)
+		}
+
+		claimRef, err := api.GetReference(claim)
+		if err != nil {
+			return nil, fmt.Errorf("Unexpected error getting claim reference: %v", err)
+		}
+
+		volumeClone.Spec.ClaimRef = claimRef
+		if boundByController {
+			ctrl.setAnnotation(&volumeClone.ObjectMeta, annBoundByController)
+		}
+
+		newVol, err := ctrl.kubeClient.Core().PersistentVolumes().Update(volumeClone)
+		if err != nil {
+			glog.V(5).Infof("Updating PersistentVolume[%s]: binding to %q failed: %v", volume.Name, claimToClaimKey(claim), err)
+			return newVol, err
+		}
+		glog.V(5).Infof("Updating PersistentVolume[%s]: bound to %q", newVol.Name, claimToClaimKey(claim))
+		return newVol, nil
+	}
+	glog.V(5).Infof("Updating PersistentVolume[%s]: already bound to %q", volume.Name, claimToClaimKey(claim))
+	return volume, nil
+}
+
+func (ctrl *PersistentVolumeController) bindClaimToVolume(claim *api.PersistentVolumeClaim, volume *api.PersistentVolume) (*api.PersistentVolumeClaim, error) {
+	glog.V(5).Infof("Updating PersistentVolumeClaim[%s]: binding to %q", claimToClaimKey(claim), volume.Name)
+
+	boundByController := false
+	// Check if the claim was already bound
+	if volume.Name != claim.Spec.VolumeName {
+		boundByController = true
+	}
+
+	// Save the claim only when any modification is neccesary.
+	if claim.Spec.VolumeName != volume.Name ||
+		!ctrl.hasAnnotation(claim.ObjectMeta, annWasEverBound) ||
+		(boundByController && !ctrl.hasAnnotation(claim.ObjectMeta, annBoundByController)) {
```

Comment:
Again, whole purpose of this check is to save the claim only if there are some changes to be done. This function should not do anything if the claim is already bound and all annotations are set.

boundByController will be `false` if the claim was bound by user, i.e. the user set `claim.Spec.VolumeName="pv1"` when creating the claim.